crates/rustc_codegen_spirv/src/builder/builder_methods.rs

```diff
@@ -540,190 +578,434 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
+                    // which were intended to operate relative to `ty`. Fall back to the standard path.
+                    trace!(
+                        "Recovery type mismatch ({}) vs ({}) and GEP indices exist, falling back",
+                        self.debug_type(ty),
+                        self.debug_type(base_pointee_ty)
                     );
                 }
+            } else {
+                // `recover_access_chain_from_offset` couldn't find a structured path for the constant offset.
+                trace!("`recover_access_chain_from_offset` returned None, falling back");
             }
         }
 
-        let result_type = self.type_ptr_to(result_pointee_type);
+        // --- End Recovery Path ---
 
-        // Check if `ptr_id` is defined by an `OpAccessChain`, and if it is,
-        // grab its base pointer and indices.
-        //
-        // FIXME(eddyb) this could get ridiculously expensive, at the very least
-        // it could use `.rev()`, hoping the base pointer was recently defined?
+        // --- Attempt GEP Merging Path ---
+
+        // Check if the base pointer `ptr` itself was the result of a previous
+        // AccessChain instruction. Merging is only attempted if the input type `ty`
+        // matches the pointer's actual underlying pointee type `original_pointee_ty`.
+        // If they differ, merging could be invalid.
         let maybe_original_access_chain = if ty == original_pointee_ty {
-            let emit = self.emit();
-            let module = emit.module_ref();
-            let func = &module.functions[emit.selected_function().unwrap()];
-            let base_ptr_and_combined_indices = func
-                .all_inst_iter()
-                .find(|inst| inst.result_id == Some(ptr_id))
-                .and_then(|ptr_def_inst| {
-                    if matches!(
-                        ptr_def_inst.class.opcode,
-                        Op::AccessChain | Op::InBoundsAccessChain
-                    ) {
-                        let base_ptr = ptr_def_inst.operands[0].unwrap_id_ref();
-                        let indices = ptr_def_inst.operands[1..]
-                            .iter()
-                            .map(|op| op.unwrap_id_ref())
-                            .collect::<Vec<_>>();
-                        Some((base_ptr, indices))
-                    } else {
-                        None
-                    }
-                });
-            base_ptr_and_combined_indices
+            // Search the current function's instructions...
+            // FIXME(eddyb) this could get ridiculously expensive, at the very least
+            // it could use `.rev()`, hoping the base pointer was recently defined?
+            let search_result = {
+                let emit = self.emit();
+                let module = emit.module_ref();
+                emit.selected_function().and_then(|func_idx| {
```

Comment:
This accounts for `selected_function()` being `None` - that should never be possible when you are in a method of a builder that will emit instructions into a selected *block*, not just function.